trusted/pal/src/pal.rs

```diff
@@ -0,0 +1,174 @@
+//! a poll based execution model driven by a signal generating loop in untrusted
+//! launcher.
+//!
+//! [Application] trait must be implemented by a concrete trusted application to
+//! receive signals from the trusted host originating from untrusted launcher.
+//!
+//! [Host] trait must be implemented by a concrete trusted host to expose its capabilities
+//! to the trusted application.
+
+use alloc::boxed::Box;
+use alloc::vec::Vec;
+use anyhow::Result;
+
+type Message = Vec<u8>;
+
+/// Represents a set of claims for the trusted host and application, provides capability
+/// sign and verify application data.
+pub trait Attestation {
+    /// TODO: Expose application identity, provenance and configuration information such
+    /// that applications can assert the properties of the peer.
+
+    /// Serializes attestation such that it can be transferred over the wire to a peer
+    /// and attested by the peer.
+    ///
+    /// # Returns
+    ///
+    /// An error if failed to serialize into a byte string, a success otherwise.
+    ///
+    /// # Note
+    ///
+    /// The serialized representation must contain enough information to
+    /// verify the attestation itself (e.g. that it has been produced by a proper root
+    /// of trust and that it represents a trusted chain of components) and to verify
+    /// signed application data (e.g. trusted applications sharing data with each other).
+    ///
+    /// For example, before two trusted application peers can start interacting they must
+    /// establish trust. The trusted application initiating the interaction will first
+    /// obtain its own attestation, sign the data it wants to share with the peer,
+    /// share serialized attestation, the data and its signature with the peer. The peer
+    /// deserializes and verifies the attestation (see `Host::verify_peer_attestation`),
+    /// uses the attestation to verify the data signatures, performs similar handshake
+    /// with the peer.
+    fn serialize(&self) -> Result<Vec<u8>>;
+
+    /// Signs a byte string using private signing key.
+    ///
+    /// # Arguments
+    ///
+    /// * `data` - The byte string to sign.
+    ///
+    /// # Returns
+    ///
+    /// An error if the attestation doesn't possess the signing capability (e.g. this
+    /// instance has been transferred over the wire and meant to be used for verification)
+    /// or if the signing failed, a success otherwise.
+    fn sign(&self, data: &[u8]) -> Result<Vec<u8>>;
+
+    /// Verifies signature of a byte string using public signing key.
+    ///
+    /// # Arguments
+    ///
+    /// * `data` - The byte string to verify signature for.
+    /// * `signature` - The signature of the byte string.
+    ///
+    /// # Returns
+    ///
+    /// An error if the signature cannot be verifed or is invalid, a success otherwise.
+    fn verify(&self, data: &[u8], signature: &[u8]) -> Result<()>;
```

Comment:
I think `SignatureVerifier` can be a separate entity, which will take a public key taken from the `Evidence` and `certificate` containing the signed data, and verify the certificate.

And `AttestationVerifier` will take an `Evidence` and reference values (or a policy) and verify it.